### ISSUE TITLE
Dynamic Phrase Builder? 

### ISSUE THREAD

**Issue by terracoda** [phetsims/a11y-research#141]
@emily-phet and I have bounced around a few ideas about how to make a simple but effect tool that helps with description design.

Our ideas have ranged from a fully editable A11y-Vew-Type interface to this simple phrase builder that uses the select element to play with phrase parameters.

I'd like to know if this little phrase builder in a very simple form would be something useful to designers and feasible to build in a short amount of time. It doesn't have to be pretty.

**I uploaded a sample static HTML file with some potential specs:**
https://github.com/phetsims/a11y-research/blob/master/tools/phrase-and-parameter-sketcher.html

**Here's a snap shot that I made in the W3 School's code builder:**
<img width="799" alt="Screen Shot 2019-06-21 at 1 18 57 PM" src="https://user-images.githubusercontent.com/1902116/60012886-5bfe9480-967d-11e9-8e11-48d300272c78.png">

**Comment by terracoda** [comment]
Tagging @emily-phet , @jessegreenberg, @zepumph, @twant, @Matthew-phet for potential feedback.

**Comment by terracoda** [comment]
This issue is related to #136

**Comment by jessegreenberg** [comment]
@terracoda, this looks really cool! Can you comment on what this will provide that is currently not supported by current design tools like spreadsheets and docs? Is it mostly to help identify the parameters of a phrase and then see the sentence in full with params filled in or are there other things too?

https://github.com/phetsims/a11y-research/blob/master/tools/phrase-and-parameter-sketcher.html gave a good sense of what you are looking for, thanks! I think it would take a few hours to make a prototype.

**Comment by terracoda** [comment]
@jessegreenberg, this doesn't add a huge amount of functionality that a table in a design document would actually achieve. However, it is surely more interactive than a table in a design doc. 

I think a tool like this might aid a designer in playing with the phrase and the parameters to carefully refine wording and perhaps splicing before implementation.

I think by being able to play with the phrase more directly might help find edge cases that require special wording or identify annoying splicing errors.

**Comment by terracoda** [comment]
@jessegreenberg, I have a feeling that it could potentially make the design and development process more efficient be reducing wording/splicing errors, and reducing the number of required updates to tables of examples or walk-throughs. 

It could aid the designer in walking through the idea more completely in a direct way. 

If the tool could actually export any useful results of the exploration then it could really reduce the need for a design er to document a robust series of examples. This would make documentation of the design more efficient and less verbose.

**Comment by terracoda** [comment]
It's taken a long time (various reasons) to get to the final phrase and parameters in the provided example.

**Comment by terracoda** [comment]
Here's the example in HTML preview:
https://htmlpreview.github.io/?https://github.com/phetsims/a11y-research/blob/master/tools/phrase-and-parameter-sketcher.html

**Comment by zepumph** [comment]
Here is a link that you can use to look at the first prototype:

https://htmlpreview.github.io/?https://github.com/phetsims/a11y-research/blob/master/js/phrase-builder/phrase-builder.html

Basic instructions:
1. Add your sentence into the first text area, with `{{templateVars}}` annotated with the double curly braces around the sides. You can only use letters inside the braces, no spaces.
2. Once complete with the sentence, you can add the options for each var in the text areas below, each option gets its own line, and if you want an option that is blank, you can just include a blank line. Click the button for each templated var to add the options to the output sentence.
3. The output sentence should have the updated options in the select box.
4. You can edit the sentence all you want, but note that if you change a variable name, or remove it (even to copy paste it somewhere else), you will lose the options that were associated with that templated var.

Let me know what you think!

**Comment by zepumph** [comment]
Also assigning @emily-phet if she wants to comment.

**Comment by terracoda** [comment]
There are some bugs, but looks promising.
I added a {{templateVars}} after the original set and I got some weird repetition and no new select box maker-thingy.
<img width="1003" alt="Screen Shot 2019-06-26 at 8 47 46 AM" src="https://user-images.githubusercontent.com/1902116/60157520-3b564c00-97ef-11e9-8402-c35cfe562e45.png">

**Comment by terracoda** [comment]
@zepumph, I'll have to do some more investigation, but I won't have time today, I don't think.

**Comment by jessegreenberg** [comment]
Really cool @zepumph, I like it a lot. Nice work!

**Comment by zepumph** [comment]
Sorry if I wasn't clear, it is designed so that each is named differently, like "templateVar1" etc.

**Comment by terracoda** [comment]
> each is named differently, like "templateVar1"

I was wondering about that. I hadn't yet tried it, but will ;-)

**Comment by zepumph** [comment]
This is becoming a priority again, as much as it is helpful to designers. I think we will work on it more this corner.

Here is a wishlist for me after looking at it again for only a few minutes:
* local storage saving so that refresh doesn't make us lose work
* "save/load" feature so that you can download work, and then load it again, or perhaps send that to someone else to load
* Instructions on top
* It would be nice if you didn't need to manually "add" options, what if each dynamic element adder could be a bit more graceful, and automatically populate the select boxes with the best available info.

**Comment by terracoda** [comment]
Other things to keep in mind:
- keep the text on the add buttons as short as possible, e.g. [[Add `fruits`]].
- The instructional part of the button label text could be put in single general location.
- Perhaps the flow could be improved with input or edit buttons?
- I agree, some option for saving the results is essential.

**Comment by zepumph** [comment]
I got through everything in the above bullets. @terracoda let me know how this looks. I made it so that much more is just done automatically. Please review.

**Comment by emily-phet** [comment]
@zepumph - where can I find the latest?

**Comment by zepumph** [comment]
http://htmlpreview.github.io/?https://github.com/phetsims/a11y-research/blob/master/js/phrase-builder/phrase-builder.html

Does this work? Right  on we don't have a better way to host it, but I will work on that.

**Comment by zepumph** [comment]
Also note I just tried opening on a mobile device and it didn't look right. We may need a few changes to be more universal.

**Comment by terracoda** [comment]
That link no longer looks like what you shared with me last on Slack.
The link loads this:
![Screen Shot 2020-04-14 at 1 36 03 PM](https://user-images.githubusercontent.com/1902116/79247455-21a4f480-7e55-11ea-9c6f-eb6a272e7618.png)

But I expected something like this:
![Screen Shot 2020-03-30 at 5 51 47 PM](https://user-images.githubusercontent.com/1902116/79247518-37b2b500-7e55-11ea-9c7d-1c23d84a6626.png)

**Comment by zepumph** [comment]
Me neither! I just made it so that we are hosting that page through github. Please review with this link:

https://phetsims.github.io/a11y-research/js/phrase-builder/

**Comment by terracoda** [comment]
This is looking awesome @zepumph.

It could be possible to have the choose from the same list of parameters more than once per sentence show. Is it be possible to do a sentence like this?

- I like to eat, eat, eat, apples and {{otherFruit}} and {{otherFruit}}.

Or is better to do this:
- I like to eat, eat, eat, apples and {{otherFruit}} and {{otherFruit01}}.

The reason I was thinking about this was because of phrases in BASE like:
- {{Yellow balloon}} has {{many more}} {{positive}} charges than {{negative}} charges.
- {{Sweater}} has {{many more}} {{negative}} charges than {{positive}} charges.

**Comment by terracoda** [comment]
@zepumph, could you replace the instructional content with a list, so the initial text reads like step-by-step instructions:

 **Dynamic Phrase Manipulator**
This phrase creator supports rapid prototyping and verification of dynamic phrases used in dynamic descriptions and responses.

To create a phrase with dynamic parts:
 1. Type in a description phrase in the text area input. 
 2. For the dynamic parts of the phrase, create uniquely named variables surrounded by curly braces, e.g. "{{templateVariable}}". 
 3. A new text area input is created for each one of these curly-braced variables. In the newly created inputs for each variable, enter the options on a new line. 
 
The output phrase will contain a select box for each variable populated with the entered options.

Refreshing the page will automatically load the most recently auto-saved phrases.

Input raw phrase with uniquely named variables within curly braces, {{exampleVariable}}:

...the rest looks good...

**Comment by zepumph** [comment]
Because of work done in https://github.com/phetsims/a11y-research/issues/161, I wanted to make sure this component still ran, so I wrote a bit of a build step for it. Please review and test by loading https://phetsims.github.io/a11y-research/js/phrase-builder/build/, which should be the exact same as the above, (before I broke it with the other issue's work).

**Comment by terracoda** [comment]
@zepumph, I tested a bit, and there are some issues on refreshing.
I added a new variable (see resetHintTrue) and it didn't get loaded in the phrase down below. And I deleted a duplicate variable and it didn't get removed from the list (see 2 lightSource variables).

And I noticed an extra curly brace down below. I couldn't find a source from my example description.

**Comment by terracoda** [comment]
A screen shot may be helpful.
![Screen Shot 2021-02-01 at 6 26 55 PM](https://user-images.githubusercontent.com/1902116/106523675-08406800-64bc-11eb-9b1a-5da4f1d3cd75.png)

**Comment by zepumph** [comment]
How's that?

**Comment by marlitas** [comment]
Meeting 11/20:
- This issue has completed. We can close.

**Comment by zepumph** [comment]
Should we delete the code? Do we want this still usable? Last I checked I couldn't rebuild it. Perhaps we should clean it up if unneeded.

**Comment by jessegreenberg** [comment]
Can you point me to the code? Is it integrated into sim code or does it all live in a11y-research?

**Comment by jessegreenberg** [comment]
Additional thoughts on why we closed this - 

We reviewed this tool recently when we investigated https://github.com/phetsims/joist/issues/941. It definitely motivated some of the thoughts we had there about creating a description design tool. We decided not to continue with creating a description design tool for now. Perhaps PhET will look into it again someday though.

We closed the issue to reflect the work status, but keeping the code around seems useful.

I found PhraseBuilder.js and will add a note pointing to this issue there.
